[JSC] Implement Math.sumPrecise()

shvaikalesh opened the pull request:
#### f49d6a41fb8d93f9c7e6758d87a06a9eea3e9698
<pre>
[JSC] Implement Math.sumPrecise()
<a href="https://bugs.webkit.org/show_bug.cgi?id=279768">https://bugs.webkit.org/show_bug.cgi?id=279768</a>
&lt;<a href="https://rdar.apple.com/131580043">rdar://131580043</a>&gt;

Reviewed by Yusuke Suzuki.

This patch implements Math.sumPrecise() proposal [1] behind a flag by importing its polyfill [2] and
rewriting it in C++, with a minor tweak of how negative zero edge case is handled (without extra loop)
and proper iterator closure.

radfordneal/xsum [3], mentioned in the proposal, didn&apos;t pass all the test262 tests.

[1]: <a href="https://github.com/tc39/proposal-math-sum">https://github.com/tc39/proposal-math-sum</a>
[2]: <a href="https://github.com/tc39/proposal-math-sum/blob/main/polyfill/polyfill.mjs">https://github.com/tc39/proposal-math-sum/blob/main/polyfill/polyfill.mjs</a>
[3]: <a href="https://gitlab.com/radfordneal/xsum">https://gitlab.com/radfordneal/xsum</a>

* JSTests/stress/math-sum-precise.js: Added.
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::MathObject::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WTF/LICENSE-SumPrecise.txt: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/sum_precise/LICENSE.txt: Added.
* Source/WTF/wtf/sum_precise/SumPrecise.cpp: Added.
(WTF::SumPrecise::twosum):
(WTF::SumPrecise::add):
(WTF::SumPrecise::compute):
* Source/WTF/wtf/sum_precise/SumPrecise.h: Added.

Canonical link: <a href="https://commits.webkit.org/283774@main">https://commits.webkit.org/283774@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04642d774ceb91ffdbe541168eaea936fabe0f15

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/67312 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/46691 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/19944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/71356 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/18448 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/54489 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/18240 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53917 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/12376 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/70379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/42886 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/58230 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/34450 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/39558 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/15629 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/16805 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/60427 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/61527 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/15970 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/73065 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/66557 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/11271 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/15303 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/61401 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/11304 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/58294 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/61482 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/9228 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/2834 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/88326 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10220 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/42496 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/15558 "Found 6 new JSC stress test failures: wasm.yaml/wasm/stress/cc-f32-kitchen-sink.js.wasm-eager, wasm.yaml/wasm/stress/f32-tuple-jsapi-exported.js.default-wasm, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-bbq, wasm.yaml/wasm/stress/js-wasm-js-varying-arities.js.wasm-no-cjit, wasm.yaml/wasm/stress/tail-call.js.wasm-bbq, wasm.yaml/wasm/stress/tail-call.js.wasm-eager (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/43573 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/44759 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/43314 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/sum_precise/LICENSE.txt:
```diff
@@ -0,0 +1,13 @@
+BSD 3-Clause License
```
If LICENSE-SumPrecise.txt exits, this duplicate file is not necessary.

Constellation commented on Source/WTF/wtf/sum_precise/SumPrecise.h:
```diff
@@ -0,0 +1,52 @@
+    double overflow = 0;
+};
+
+void add(State&, double);
```
Can you make it member function of `State`?

Constellation commented on Source/WTF/wtf/sum_precise/SumPrecise.h:
```diff
@@ -0,0 +1,52 @@
+};
+
+void add(State&, double);
+double compute(State&);
```
Ditto.

Constellation commented on Source/WTF/wtf/sum_precise/SumPrecise.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace SumPrecise {
+
+struct State {
```
Let's make it `class`.

Constellation commented on Source/WTF/wtf/sum_precise/SumPrecise.h:
```diff
@@ -0,0 +1,52 @@
+struct State {
+    bool everyValueIsNegativeZero = true;
+    Vector<double> partials;
+    double overflow = 0;
```
Can you make them `private` `m_everyValueIsNegativeZero` etc?

Constellation reviewed (commented):
Can you put SumPrecise.cpp / SumPrecise.h under wtf/ instead of sub directory given that it is not importing third party files directly?

Constellation commented on Source/WTF/wtf/sum_precise/SumPrecise.h:
```diff
@@ -0,0 +1,52 @@
+ * License header from Math.sumPrecise() polyfill
+ *    https://github.com/tc39/proposal-math-sum/blob/main/polyfill/LICENSE
+ */
+
```
Or, if it is just one cpp / h file, you can just paste license here as a comment.

kmiller68 replied:
I'm not sure the license is needed since this is a re-implementation from JS -> C++? Although, regardless, credit is reasonable/respectful.

Constellation reviewed (approved):
r=me

Commit:
[JSC] Implement Math.sumPrecise()
https://bugs.webkit.org/show_bug.cgi?id=279768
<rdar://131580043>

Reviewed by Yusuke Suzuki.

This patch implements Math.sumPrecise() proposal [1] behind a flag by importing its polyfill [2] and
rewriting it in C++, with a minor tweak of how negative zero edge case is handled (without extra loop)
and proper iterator closure.

radfordneal/xsum [3], mentioned in the proposal, didn't pass all the test262 tests.

[1]: https://github.com/tc39/proposal-math-sum
[2]: https://github.com/tc39/proposal-math-sum/blob/main/polyfill/polyfill.mjs
[3]: https://gitlab.com/radfordneal/xsum

* JSTests/stress/math-sum-precise.js: Added.
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::MathObject::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/WTF/LICENSE-SumPrecise.txt: Added.
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/sum_precise/LICENSE.txt: Added.
* Source/WTF/wtf/sum_precise/SumPrecise.cpp: Added.
(WTF::SumPrecise::twosum):
(WTF::SumPrecise::add):
(WTF::SumPrecise::compute):
* Source/WTF/wtf/sum_precise/SumPrecise.h: Added.

Canonical link: https://commits.webkit.org/283774@main
